File: apps/lexicoff/src/lib/romanizations.ts
import profiles from '../../../../shared/romanization-profiles.json';

// Keep normalization aligned with pipeline/romanizations.py and shared fixtures.
export function romanizationKeys(
	lang: string,
	reading: string
): { key: string; quality: number }[] {
	if (!Object.hasOwn(profiles, lang)) return [];
	const strict = reading.trim().toLowerCase().normalize('NFC');
	if (!strict) return [];
	const mapping = profiles[lang as keyof typeof profiles] as Record<string, string>;
	const letters = Object.keys(mapping).sort((a, b) => b.length - a.length);
	let loose = strict;
	let pattern: RegExp | undefined;
	if (letters.length) {
		const source = letters.map((s) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')).join('|');
		pattern = new RegExp(source, 'gu');
		loose = loose.replace(pattern, (match) => mapping[match]);
	}
	loose = loose.normalize('NFD').replace(/\p{Mn}/gu, '');
	if (pattern) loose = loose.replace(pattern, (match) => mapping[match]);
	loose = loose
		.replace(/[’'ʼʹʺ]/gu, '')
		.replace(/[\s-]+/gu, ' ')
		.trim();
	return [{ key: loose || strict, quality: 5 }];
}

File: pipeline/romanizations.py
"""Explicit headword readings only; no inferred vowels or paradigm expansion."""
import json
from pathlib import Path
import re
import unicodedata

PROFILES = json.loads((Path(__file__).resolve().parent.parent / 'shared/romanization-profiles.json').read_text())
VERSION = '1'


def romanization_keys(lang, reading):
    if lang not in PROFILES:
        return []
    strict = unicodedata.normalize('NFC', reading.strip().lower())
    if not strict:
        return []
    mapping = PROFILES[lang]
    loose = strict
    if mapping:
        pattern = '|'.join(re.escape(key) for key in sorted(mapping, key=len, reverse=True))
        loose = re.sub(pattern, lambda match: mapping[match[0]], loose)
    loose = ''.join(c for c in unicodedata.normalize('NFD', loose) if unicodedata.category(c) != 'Mn')
    if mapping:
        loose = re.sub(pattern, lambda match: mapping[match[0]], loose)
    loose = re.sub(r"[’'ʼʹʺ]", '', loose)
    loose = re.sub(r'[\s-]+', ' ', loose).strip()
    # Display retains the original spelling; index only one normalized key.
    return [(loose or strict, 5)]


def extract_romanizations(entry, clean_text):
    if entry.lang_code not in PROFILES:
        return []
    result = []
    for form in entry.forms:
        # Table cells and alternative native spellings aren't headword readings.
        if form.source:
            continue
        raw = form.form if form.tags & {'romanization', 'transliteration'} else (
            form.roman if form.form == entry.word else None
        )
        value = clean_text(raw or '')
        if not value or value in {'-', '—', '–', '?'} or any(c in value for c in '/,;()[]{}<>'):
            continue
        # Only Latin-script readings (plus marks and ordinary separators).
        if not any('LATIN' in unicodedata.name(c, '') for c in value):
            continue
        if any(c.isalpha() and c not in 'ʿʾʹʺʼ' and 'LATIN' not in unicodedata.name(c, '') for c in value):
            continue
        if value not in result:
            result.append(value)
    return result
